flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/InputProcessorUtil.java

```diff
@@ -71,108 +69,55 @@ public static CheckpointedInputGate createCheckpointedInputGate(
 			String taskName,
 			List<IndexedInputGate>... inputGates) {
 
-		IntStream numberOfInputChannelsPerGate =
-			Arrays
-				.stream(inputGates)
-				.flatMap(collection -> collection.stream())
-				.sorted(Comparator.comparingInt(IndexedInputGate::getGateIndex))
-				.mapToInt(InputGate::getNumberOfInputChannels);
-
-		Map<InputGate, Integer> inputGateToChannelIndexOffset = generateInputGateToChannelIndexOffsetMap(unionedInputGates);
-		// Note that numberOfInputChannelsPerGate and inputGateToChannelIndexOffset have a bit different
-		// indexing and purposes.
-		//
-		// The numberOfInputChannelsPerGate is indexed based on flattened input gates, and sorted based on GateIndex,
-		// so that it can be used in combination with InputChannelInfo class.
-		//
-		// The inputGateToChannelIndexOffset is based upon unioned input gates and it's use for translating channel
-		// indexes from perspective of UnionInputGate to perspective of SingleInputGate.
-
+		IndexedInputGate[] sortedInputGates = Arrays.stream(inputGates)
+			.flatMap(Collection::stream)
+			.sorted(Comparator.comparing(IndexedInputGate::getGateIndex))
+			.toArray(IndexedInputGate[]::new);
 		CheckpointBarrierHandler barrierHandler = createCheckpointBarrierHandler(
 			config,
-			numberOfInputChannelsPerGate,
+			sortedInputGates,
 			checkpointCoordinator,
 			taskName,
-			generateChannelIndexToInputGateMap(unionedInputGates),
-			inputGateToChannelIndexOffset,
 			toNotifyOnCheckpoint);
 		registerCheckpointMetrics(taskIOMetricGroup, barrierHandler);
 
+		InputGate[] unionedInputGates = Arrays.stream(inputGates)
```

Comment:
It's a bit confusing that above we have `sortedInputGates` and those input gates here are not sorted. It means we end up with a confusing state, where `CheckpointBarrierHandler#inputGates` can be accessed via `inputGateIndex` while `UnionInputGate#inputGates` can not be.

I understand why is it so, first one is flattened structure of all input gates, while the other has only a subset of gates. Maybe we can keep it as it is for now, as this commit is already simplifying things, but maybe we should replace `UnionInputGate#inputGates` array with a map?